src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1272,30 +1279,117 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
   8. Need to figure out a way to get all versions to be deleted in case of versioned objects when a version is not specified.
   9. If the object is not in object directory call next->delete_obj
   */
-  rgw::d4n::CacheObj obj = rgw::d4n::CacheObj{ // TODO: Add logic to ObjectDirectory del method to also delete all blocks belonging to that object
-			     .objName = source->get_key().get_oid(),
-			     .bucketName = source->get_bucket()->get_name()
-			   };
 
-  int ret = -1;
-  if ((ret = source->driver->get_obj_dir()->del(dpp, &obj, y)) < 0) 
-    ldpp_dout(dpp, 0) << "D4NFilterObject::" << __func__ << "(): ObjectDirectory del() method failed, ret=" << ret << dendl;
+  rgw::sal::Attrs attrs;
+  std::string head_oid_in_cache;
+  if (!source->check_head_exists_in_cache_get_oid(dpp, head_oid_in_cache, attrs, y)) {
+    ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): calling next delete_obj" << dendl;
+    return next->delete_obj(dpp, y, flags);
+  } else {
+    rgw::d4n::CacheObj obj = rgw::d4n::CacheObj{ 
+			       .objName = source->get_name(),
+			       .bucketName = source->get_bucket()->get_name()
+			     };
 
-  Attrs::iterator attrs;
-  Attrs currentattrs = source->get_attrs();
-  std::vector<std::string> currentFields;
-  
-  /* Extract fields from current attrs */
-  for (attrs = currentattrs.begin(); attrs != currentattrs.end(); ++attrs) {
-    currentFields.push_back(attrs->first);
-  }
+    int ret = -1;
+    auto rgwObj = source->get_obj();
+    ret = source->get_obj_attrs(y, dpp, &rgwObj);
```

Comment:
why is this call needed?

Reply:
To populate the object's size since `source->get_obj_size()` is called later. If I don't call this method first, the value is 0.

Reply:
You can check attrs returned by check_head_oid_exists_in_cache and get the object size using user.rgw.object_size. This way double calls to check_head_oid_exists_in_cache can be avoided

Reply:
I removed this and I'm using the attr returned from `check_head_exists_in_cache_get_oid`